Fix resource queue slot leaks involving self-deadlocks and termination

soumyadeep2007 opened the pull request:
This PR consists of:

1) Tests adding coverage and documentation for resource queue behavior
2) The fix for the statement leak involving self-deadlocks
3) The fix for the statement leak involving termination
4) Logging improvements

Please review commits individually

Dev-pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/resq_self

jimmyyih commented on src/backend/utils/resscheduler/resqueue.c:
```diff
@@ -1615,6 +1627,12 @@ ResCheckSelfDeadLock(LOCK *lock, PROCLOCK *proclock, ResPortalIncrement *increme
 						memoryThesholdOvercommitted = true;
+						ereport(LOG,
+								(errmsg("memory threshold overcommitted"),
+									errdetail("total cost %lf exceeds limit %f for resource queue id: %u",
```
Should this say "total memory %lf exceeds limit"?

soumyadeep2007 replied:
Yes indeed. Will make that change.

jimmyyih commented on src/test/isolation2/expected/resource_queue_role.out:
```diff
@@ -0,0 +1,136 @@
+1:DECLARE c2 CURSOR FOR SELECT 1;
+DECLARE CURSOR
+
+-- We should see 1 lock each on each queue, with 1 holder, with 1 active
+-- statement in queue. The user in gp_locks_on_resqueue will be the session_user
+-- as opposed to the current_user 'rq_role_test1' or 'rq_role_test2' (which we
+-- use to find the destination queue)
```
For this and the other tests below, I got a bit confused on this line:
> The user in gp_locks_on_resqueue will be the session_user as opposed to the current_user 'rq_role_test1' or 'rq_role_test2'

It took me some time to realize the comment was being represented by the `lorusename=session_user`.  It didn't help that the comment doesn't have the correct role names so I got them confused with the `lorrsqname` in the query output.  That being said, I recommend changing it to something like this:
> The lorusename in gp_locks_on_resqueue will be the session_user as opposed to the current_user 'rq_role_test_role1' or 'rq_role_test_role2'

Or even changing the role names to `rq_role_test1_user` and `rq_role_test2_user`.

soumyadeep2007 replied:
> The lorusename in gp_locks_on_resqueue will be the session_user as opposed to the current_user 'rq_role_test_role1' or 'rq_role_test_role2'

Sounds good.

soumyadeep2007 commented on src/backend/utils/resscheduler/resqueue.c:
```diff
@@ -510,6 +510,22 @@ ResLockAcquire(LOCKTAG *locktag, ResPortalIncrement *incrementSet)
+		 */
+		if (ResCheckSelfDeadLock(lock, proclock, incrementSet))
+		{
+			LWLockRelease(partitionLock);
+			LWLockRelease(ResQueueLock);
```
Yes, I'll make that change. Let's be consistent, thanks for catching.

huansong commented on src/test/isolation2/expected/resource_queue_self_deadlock.out:
```diff
@@ -0,0 +1,78 @@
+
+2: SET ROLE r_self;
+SET
+-- Will wait for a slot in the queue.
+2&: SELECT 33176;  <waiting ...>
+
+-- Now inflate the active statement count for the queue.
+0<:  <... completed>
+ gp_wait_until_triggered_fault 
+-------------------------------
+ Success:                      
+(1 row)
+0: ALTER RESOURCE QUEUE rq_self WITH (active_statements = 5);
+ALTER QUEUE
+
+-- Cancellation of the backend should not lead to us doing an extra grant to the
+-- process undergoing self-deadlock. Doing so, leads to an assertion failure:
+-- FailedAssertion("!(lock->nGranted <= lock->nRequested)", File: "resqueue.c", Line: 1086)
+-- ResGrantLock resqueue.c:1086
+-- ResProcLockRemoveSelfAndWakeup resqueue.c:1361
+-- ResCleanUpLock resqueue.c:1215
+-- ResRemoveFromWaitQueue resqueue.c:1473
+-- ResLockWaitCancel proc.c:2102
+-- ResLockPortal resscheduler.c:695
+-- ...
+3: SELECT pg_cancel_backend(pid) FROM pg_stat_activity WHERE query='SELECT 33176;';
+ pg_cancel_backend 
+-------------------
+ t                 
+(1 row)
+2<:  <... completed>
+ERROR:  canceling statement due to user request
```
Q: does session 2 immediately proceed once we inflate the statement limit? Mainly thinking if there's a flakiness if session 2 completes the query before we cancel it. Or is it blocked by something else?

soumyadeep2007 replied:
Session 2 before the cancel_backend call is blocked in ResProcSleep() -> WaitLatch(). Then it gets the cancellation and can proceed after that. The `2<:` should block until the statement finishes erroring out.

PS: The alter queue operation does nothing apart from **updating the value of the limit of the queue**. That means that it doesn't do anything to change the **current consumption** of the queue.

zxuejing commented:
Hi, I reproduce this issue on 6_stable use the following testcases,
but I can not reproduce this issue on master use the same testcases.
Test case:
```
session 1:
      create resource queue myqueue WITH (ACTIVE_STATEMENTS=1);
      create user u1  resource queue myqueue;
      set role u1;
      begin;
      declare c1 cursor for select * from pg_class;
session 2:
      set role u1;
      select 1; --waiting

gdb attach pid of session 1 and break point in line ResProcWakeup on ResProcLockRemoveSelfAndWakeup. 
 session 1:
      close c1;
      -- this is stopped in above breakpoint

session3:
      select pg_terminate_backend(the pid of session 2);

      continue gdb of session 1;

session 3:
      select * from gp_toolkit.gp_resqueue_status; -- every thing is ok in master.
```

I want ask some questions:
1、Do we have this issue on master, if yes, how to reproduce it.
2、There are many bugs we fixed in 6_stable and do not backported to master.
      Do we need to fix this on master?

zxuejing commented:
Hi, I reproduce this issue on 6_stable use the following testcases,
but I can not reproduce this issue on master use the same testcases.
Test case:
```
session 1:
      create resource queue myqueue WITH (ACTIVE_STATEMENTS=1);
      create user u1  resource queue myqueue;
      set role u1;
      begin;
      declare c1 cursor for select * from pg_class;
session 2:
      set role u1;
      select 1; --waiting

gdb attach pid of session 1 and break point in line ResProcWakeup on ResProcLockRemoveSelfAndWakeup. 
 session 1:
      close c1;
      -- this is stopped in above breakpoint

session3:
      select pg_terminate_backend(the pid of session 2);

      continue gdb of session 1;

session 3:
      select * from gp_toolkit.gp_resqueue_status; -- every thing is ok in master.
```

I want ask some questions:
1、Do we have this issue on master, if yes, how to reproduce it.
2、There are many bugs we fixed in 6_stable and do not backported to master.
      Do we need to fix this on master?

kainwen commented:
@zxuejing 
> There are many bugs we fixed in 6_stable and do not backported to master.

Can you give more details on this? I believe any bug fix should 1st happen on main branch and then backport to 6X.

zxuejing commented:
> Can you give more details on this? I believe any bug fix should 1st happen on main branch and then backport to 6X.
@kainwen I checked some pr and only this pr #14351 was not modified in the master.
Bugs are fixed in both 6X_stable and master.

soumyadeep2007 commented:
> I want ask some questions:
1、Do we have this issue on master, if yes, how to reproduce it.
2、There are many bugs we fixed in 6_stable and do not backported to master.
Do we need to fix this on master?

Yes absolutely, it certainly does exist. You can do 2 things:

1. gdb itself may be intercepting the signals. It is best to attach after inducing a sleep.

 Apply 
[repro2.patch.txt](https://github.com/greenplum-db/gpdb/files/14160036/repro2.patch.txt) on top of main and then:

```
-- fooqueue1 has 1 active statement slot.

Session1:
postgres=# set role foorole1;
SET
postgres=> begin;
BEGIN
postgres=> declare c1 cursor for select 1;
DECLARE CURSOR

Session2:
postgres=# set role foorole1;
SET
postgres=> SELECT 1;
-- ^ HANGS waiting for slot

Session1:
CLOSE c1;
-- ^ will take 20s, due to patch's sleep

Session3 (run this within the above 20s window):
SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE query='SELECT 1;';

Session2:
FATAL:  terminating connection due to administrator command
FATAL:  Unexpected internal error (proc.c:996)
DETAIL:  FailedAssertion("!(SHMQueueEmpty(&(MyProc->myProcLocks[i])))", File: "proc.c", Line: 996)
HINT:  Process 3293408 will wait for gp_debug_linger=120 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
```

2. If you run the isolation2 test resource_queue_terminate.sql without the fix (scenario 5), it should repro. Also, you should be able to run the steps manually and that should work too.



> @zxuejing
> 
> > There are many bugs we fixed in 6_stable and do not backported to master.
> 
> Can you give more details on this? I believe any bug fix should 1st happen on main branch and then backport to 6X.

Yes, can you point us to these fixes? Definitely needs to go into 7X if it hasn't already.

soumyadeep2007 commented on src/backend/utils/resscheduler/resqueue.c:
```diff
@@ -1224,6 +1407,45 @@ ResProcLockRemoveSelfAndWakeup(LOCK *lock)
+
+		LWLockAcquire(ResQueueLock, LW_SHARED);
+		resPortalIncrement = ResIncrementFind(&portalTag);
+		LWLockRelease(ResQueueLock);
```
> Is it possible that if resPortalIncrement is null, But someone generate resPortalIncrement after we release the lock？

Great question! Its not actually possible for that to happen because for an external grant to happen, this process must be on the wait queue of the LOCK object. And, by this time, `AbortOutOfAnyTransaction() -> .. -> ResLockWaitCancel()` would already have been called and this process itself would have taken itself out of the wait queue. So, effectively at this point, no concurrent external grant can happen.

yaowangm commented:
@zxuejing https://github.com/greenplum-db/gpdb/pull/14351 is not a fix but kind of debugging code for the RQ leak issue. Once the issue is completely resolved, we may remove the code.

huansong commented on src/backend/utils/mmgr/portalmem.c:
```diff
@@ -1194,7 +1197,14 @@ AtExitCleanup_ResPortals(void)
 		Portal		portal = hentry->portal;
 
-		if (IsResQueueLockedForPortal(portal))
+		/*
+		 * Note: There is no real need to call ResPortalHasDanglingIncrement()
+		 * here. Only persisted holdable portals are cleaned up here and they
+		 * should already have hasResQueueLock=true.
+		 *
+		 * But we still do so out of paranoia/future-proofing.
+		 */
+		if (IsResQueueLockedForPortal(portal) || ResPortalHasDanglingIncrement(portal))
 			ResUnLockPortal(portal);
```
I saw that there's a few other places that checks if `IsResQueueLockedForPortal` and then tries to `ResUnLockPortal`, like in `PortalCleanup`. I have not thought the implication fully (like what's the difference between `PortalDrop` and `PortalCleanup`?), but for more future-proofing, should we also checks `ResPortalHasDanglingIncrement` there? Or, does it make sense to just make `ResPortalHasDanglingIncrement` part of `IsResQueueLockedForPortal` so we will always take the dangling increment into account when checking the latter? Probably an over-kill but just to poke thoughts.

huansong replied:
Makes sense. The current places should suffice. Thanks for the explanation!

huansong reviewed (approved):
LGTM!
